activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb

```diff
@@ -128,7 +128,14 @@ def foreign_key_options
       def columns
         result = [["#{name}_id", type, options]]
         if polymorphic
-          result.unshift(["#{name}_type", :string, polymorphic_options])
+          options = polymorphic_options
+          if index
+            indexable_string_limit = Base.connection.max_indexable_string_limit
```

Comment:
Unfortunately, there is no way. Because the charset of newly created column depends on the default charset of its table or database. We need to ask to MySQL to know the default charset of table or database. (`max_indexable_string_limit` is depends on charset in MySQL.)

By the way, it seems that this code will not be bottle neck. Because this code will be ran only when we migrate database schema. We have more heavy processes on migrating.